Do not ask for storage permissions when it's a fresh installation

grzesiek2010 opened the pull request:
Closes #4288

#### What has been done to verify that this works as intended?
I tested the behavior.

#### Why is this the best possible solution? Were any other approaches considered?
I decided to call `requestStoragePermissions()` like before but add one trick which is here:
https://github.com/getodk/collect/pull/4297/files#diff-fe2bc5034d6dfe946ff681f198d0552c83038caed992865f5e7ce030447a276bR98

Another option would be to check if we need storage permission before calling `requestStoragePermissions()` but that would require adding more if/else statements and the code would be more cluttered.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Fresh installs should not ask for storage permission that's the only change. In order to confirm that everything works fine it would be good to go through some functionalities where normally we would be asked for permissions.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

grzesiek2010 committed:
Implemented AppStateProvider

grzesiek2010 committed:
Enable scoped storage for fresh installs

grzesiek2010 committed:
Don't ask for storage permission if scoped storage is used

grzesiek2010 committed:
Use injected permissionUtils

grzesiek2010 committed:
Pass storageStateProvider to PermissionUtils

grzesiek2010 committed:
Use storageStateProvider object to determine if should ask for storage permission

grzesiek2010 committed:
areStoragePermissionsGranted() shoul not be static

grzesiek2010 committed:
Fixed areStoragePermissionsGranted()

grzesiek2010 committed:
Removed redundant code from providers

grzesiek2010 committed:
Fixed tests

lognaturel reviewed (commented):
Would be great to get some testing. Maybe just a small test for `areStoragePermissionsGranted` that verifies that `true` is always returned if scoped storage is enabled?

I like what you've done with `PermissionUtils`. I'm wondering whether it's also a good opportunity to rename it to maybe `PermissionsRequester` or `PermissionsProvider`?

grzesiek2010 commented:
> Would be great to get some testing

I thought the same at the end of my day yesterday. I'll try to add something valuable.

grzesiek2010 committed:
Call Dexter.withContext instead of deprecated withActivity

grzesiek2010 committed:
Removed redundant finishAllActivities() method from PermissionUtils

grzesiek2010 committed:
Got rid of static methods in PermissionUtils class

grzesiek2010 committed:
Implemented PermissionsChecker

grzesiek2010 committed:
Do not pass dialogTheme to PermissionUtils constructor

grzesiek2010 committed:
Moved permissionUtils class to permissions package

grzesiek2010 committed:
Renamed PermissionUtils -> PermissionsProvider

grzesiek2010 committed:
Removed duplicated AlwaysGrantStoragePermissionsPermissionsProvider class

grzesiek2010 committed:
Added tests

grzesiek2010 committed:
Reuse PermissionsProvider injected to CollectAbstractActivity

grzesiek2010 commented:
> Would be great to get some testing. Maybe just a small test for areStoragePermissionsGranted that verifies that true is always returned if scoped storage is enabled?

I like what you've done with PermissionUtils. I'm wondering whether it's also a good opportunity to rename it to maybe PermissionsRequester or PermissionsProvider?

It's ready, I added some tests and implemented code refactoring.

lognaturel commented on collect_app/src/test/java/org/odk/collect/android/permissions/PermissionsProviderTest.java:
```diff
@@ -0,0 +1,141 @@
+    @Test
+    public void whenStoragePermissionsNotGrantedAndScopedStorageUsed_shouldAreStoragePermissionsGrantedReturnTrue() {
+        when(permissionsChecker.isPermissionGranted(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE)).thenReturn(false);
+        when(storageStateProvider.isScopedStorageUsed()).thenReturn(false);
```
This should be `true` so that the last assertion is `true`!

grzesiek2010 replied:
Ahh good catch, thanks!

lognaturel reviewed (changes requested):
Looks good other than the main test... 🙃 Feel free to mark as `needs testing` once that's fixed and conflicts are resolved.

grzesiek2010 committed:
Fixed bugs in tests

kkrawczyk123 commented:
@grzesiek2010 Not asking for storage permissions with a fresh install causes crashes when try to use widgets with adding files from device's storage - image, video, audio, annotate and file widget as those widgets need storage permission enabled. I can enable it in settings/ app permissions and then it works well but none of the users would actually do that. If you really want to remove asking for permissions right after installation perhaps adding another permission dialog after using the "Choose" buttons should be considered. 

@getodk-bot unlabel "needs testing"

grzesiek2010 commented:
> Not asking for storage permissions with a fresh install causes crashes when try to use widgets with adding files from device's storage - image, video, audio, annotate and file widget as those widgets need storage permission enabled.

Does that depend on Android version or the source of files? I tested Api 29 and yeah the behavior is not consistent sometimes I get a message that I can't access a file that should be accessible but I wasn't able to reproduce a crash.

mmarciniak90 commented:
We noticed slight differences depending on the version, e.g. Android 10 does not crash, it just shows the toast `Error Occured`. The easiest way to reproduce is to add an attachment from the IMAGES / AUDIO / VIDEOS directory.

<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/105030786-9a745500-5a54-11eb-9f14-8244a687f07e.png">

grzesiek2010 commented:
Ok everything seems to come down to Storage Access Framework:
https://developer.android.com/guide/topics/providers/document-provider
https://developer.android.com/training/data-storage/shared/documents-files

and the question I asked here: https://github.com/getodk/collect/pull/4351#issuecomment-761998944

looks like we not only don't need that nasty code (which I described in the question attached above) to retrieve files but we even have to stop using it.

grzesiek2010 committed:
Use Storage Access Framework to access files in MediaLoadingTask

grzesiek2010 committed:
All widgets that use files should use the same code to load them

grzesiek2010 committed:
Removed unused code

grzesiek2010 committed:
Fixed custom splash screen image

grzesiek2010 committed:
Remove unused code

grzesiek2010 commented:
Ok I implemented some changes. @mmarciniak90 @kkrawczyk123 please test:
- capturing and attaching files in media widgets
- importing settings using a qr code from saved image (not scanning)
- setting custom splash screen images

did I miss something where we attach files?

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/dao/helpers/ContentResolverHelper.java:
```diff
@@ -76,20 +73,8 @@ public static String getFormPath(Uri uri) {
-        }
+    public static String getFileExtensionFromUri(Uri fileUri) {
+        String mimeType = getContentResolver().getType(fileUri);
+        return mimeType.substring(mimeType.lastIndexOf("/") + 1);
```
This happens to work for some common types like `image/jpeg` but the subtype name is not guaranteed to be the extension. For example the SVG MIME type is `image/svg+xml` but its extension is `.svg`. Or `audio/amr-wb` and `.amr`. I think you have to use [MimeTypeMap#getExtensionFromMimeType](https://developer.android.com/reference/android/webkit/MimeTypeMap#getExtensionFromMimeType(java.lang.String)).

grzesiek2010 replied:
Generally you are right and I tried it but in some cases (which seem pretty obvious) it returns null those cases are like:
- audio/mp3
- text/csv

Here https://stackoverflow.com/questions/12473851/how-i-can-get-the-mime-type-of-a-file-having-its-uri/12473985 where that solution is marked as accepted people also complain about the same.

Maybe I should use what you mentioned first and then in case of null my: `mimeType.substring(mimeType.lastIndexOf("/") + 1);`

or better yet (as a second step) I can read filename that should contain file extension:
```
        Cursor cursor = getContentResolver().query(fileUri, null, null, null, null);
        if (cursor != null && cursor.moveToFirst()) {
            name = cursor.getString(cursor.getColumnIndex(MediaStore.MediaColumns.DISPLAY_NAME));
        }
        String extension = name != null ? name.substring(name.lastIndexOf(".") + 1) : "";
```

this is what this library does https://github.com/anggrayudi/SimpleStorage. I know it's new and didn't gain traction yet but i ran into it through this topic https://stackoverflow.com/questions/63080879/get-file-extension-using-android-storage-access-framework.

What do you think?

lognaturel replied:
ANDROID! Why must it always make simple things hard. Good catch on `MimeTypeMap#getExtensionFromMimeType` not actually working all the time.

Doing that first and then falling back seems good. I think you might actually want to have both fallbacks since neither is perfect -- `MediaStore.MediaColumns.DISPLAY_NAME` is not necessarily a file name. Maybe first use `MimeTypeMap#getExtensionFromMimeType`, then try the display column and then use the subtype?

grzesiek2010 replied:
Ok I did it I even added another fourth aproach since I noticed that in case of selfie video only that last approach works well :scream:

lognaturel replied:
Yikes! Is that maybe because selfie video gives back a file URI (not a content URI) without a MIME type set? Maybe that's worth looking at again after this PR.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java:
```diff
@@ -87,12 +86,6 @@ public static void releaseDatabaseHelper() {
     @Override
     public boolean onCreate() {
-
-        if (!areStoragePermissionsGranted(getContext())) {
```
I think this can all go away once we don't support files in `/sdcard/odk/` anymore but at the moment, removing it could result in a crash if someone's migration failed, they haven't yet granted permissions, and they try to access forms from an external app, right?

grzesiek2010 replied:
I tested it using collectTester app and:
- if you open list of blank form everything is fine, you are asked to grant permissions
- if you open list of saved forms **as above**
- if you open list of forms to submit **as above**
- if you open list of sent forms **as above**
- if you download forms passing ids **as above**
- if you upload forms passing ids **as above**

It only crashes if an external app tries to get list of forms or instances.
The problem was with injecting objects in content providers which are initialized before other app components but finally I fixed this issue thanks to https://stackoverflow.com/questions/23521083/inject-database-in-a-contentprovider-with-dagger
https://github.com/getodk/collect/pull/4297/commits/0c1ff84374c4ca7dc7c83067cea384473fcb1767

lognaturel replied:
Makes sense!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/provider/InstanceProvider.java:
```diff
@@ -87,11 +86,6 @@ public static void releaseDatabaseHelper() {
 
     @Override
     public boolean onCreate() {
-        if (!areStoragePermissionsGranted(getContext())) {
```
Same question as for `FormsProvider`.

grzesiek2010 replied:
As above in case of FormsProvider

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/MediaUtils.java:
```diff
@@ -82,362 +60,8 @@ public void openFile(Context context, File file) {
     }
 
-    public String getMimeType(String url) {
+    private String getMimeType(String url) {
```
This seems unnecessary and likely to fail. It should be possible to use [`FileProvider#getType`](https://developer.android.com/reference/androidx/core/content/FileProvider#getType(android.net.Uri)).

grzesiek2010 replied:
Done.

grzesiek2010 replied:
I'm rolling back this. I tested that way and it doesn't work in all cases but the old one does at leas for the cases I tested. Oh Android you are so funny! 
I added some fallbacks just in case like in getFileExtensionFromUri() above.

lognaturel commented:
> @mmarciniak90 @kkrawczyk123 please test:

The really important things to check are around edge cases for file types, file extensions and file sources. For example, what if you have a jpg with the correct MIME type but no extension? What if you have a text file with extension `.fake` and the correct MIME type? What if you have files coming from Google Drive? Checking that for one place where files are attached should be sufficient.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/AudioWidget.java:
```diff
@@ -144,40 +144,25 @@ public IAnswerData getAnswer() {
         }
     }
 
-    /**
```
`FormSaveViewModel.createAnswerFile` returns a `LiveData<Result<String>>` used to set the data from a file name. Have you tried this change with internal recording?

grzesiek2010 replied:
I tested the demo form with internal recording and everything seems fine.

lognaturel replied:
Interesting! @seadowg, would be good for you to at least see this change happening since it's to very recent audio code.

seadowg replied:
Looks like @lognaturel is right as `AudioRecordingTest` and `AudioWidgetTest` are failing. From debugging, it looks like `AudioWidget#setData` is still passed a `String`.

lognaturel replied:
Do you remember whether there was a specific reason you wanted to use a `String`, @seadowg? Looks like using a `LiveData<Result<File>>` works fine but want to make sure there isn't a strong reason against it.

grzesiek2010 replied:
Yeah I have just turned it into files so please answer if there was any reason.

seadowg replied:
I don't **think** it needs to be a `String`. It was just because it didn't need to be a file as the the other side could just use `getAnswerFile` with the name (the `String`) to access the actual file. I think if `File` is more consistent we can go with that!

grzesiek2010 committed:
Protect the app from crashes when ContentProviders are used from external apps

grzesiek2010 committed:
Fixed getMimeType() method

grzesiek2010 committed:
Removed unused methods from FileUtil class

grzesiek2010 committed:
Fixed bugs

grzesiek2010 committed:
Improved VideoWidget and opening files

grzesiek2010 committed:
Improved getFileExtensionFromUri() method

grzesiek2010 committed:
Fixed bug in MediaUtils#openFile() method

grzesiek2010 committed:
Improved getFileExtensionFromUri() method

grzesiek2010 committed:
Fixed getMimeType()

grzesiek2010 committed:
Fixed pmd

grzesiek2010 committed:
Improved saveAnswerFileFromUri method

grzesiek2010 committed:
Improved getFileExtensionFromUri() methdo

grzesiek2010 committed:
Fixed pmd

grzesiek2010 committed:
Fixed ArbitraryFileWidgetTest

grzesiek2010 committed:
Fixed VideoWidgetTest

grzesiek2010 committed:
Fixed AudioWidget

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/dao/helpers/ContentResolverHelper.java:
```diff
@@ -76,20 +79,43 @@ public static String getFormPath(Uri uri) {
+    public static String getFileExtensionFromUri(Uri fileUri) {
+        String mimeType = getContentResolver().getType(fileUri);
+
+        String extension = fileUri.getScheme() != null && fileUri.getScheme().equals(ContentResolver.SCHEME_CONTENT)
```
Wouldn't we always be dealing with content URIs here? In what cases would we have file URIs?

lognaturel replied:
Or was that the selfie video case again?

grzesiek2010 replied:
This was the fix added for a crash some time ago https://github.com/getodk/collect/commit/f98b98dddc3cc9af8e1db5144f4c9c7313854df1 so I wanted to keep it just in case seeing how other cases might be tricky.

lognaturel replied:
Got it. Would be great to follow up with #2834 because this method has gone through lots of big changes during this PR that probably would have been avoided with tests.

grzesiek2010 commented:
@lognaturel Generally I think it's ready for testing unless you see something important.

lognaturel reviewed (approved):
This has turned into an adventure! Good to see so much code going away.

Lots of little things to check around the media selection questions like are temporary files cleaned up and all the [cases around MIME types and file extensions I listed earlier](https://github.com/getodk/collect/pull/4297#issuecomment-763264011).

grzesiek2010 committed:
Check Uri permission before accesing it

grzesiek2010 committed:
Check Uri permission before accesing it in ODKView

grzesiek2010 committed:
Improved tests for intent group

grzesiek2010 committed:
Moved the new message about not granted file permission to strings

grzesiek2010 committed:
Added tests to PermissionsProviderTest

grzesiek2010 commented:
I added some improvements to handle `SecurityException` which is thrown when read permission is no granted to the returned uri. I was even able to reproduce it on an emulator Api 25 where I can't record video because apparently the default app doesn't grant read permission to returned uri. Please keep an eye on it during testing.

lognaturel commented:
> I was even able to reproduce it on an emulator Api 25

Nice. But does this mean that recording video with that default app would never work? I wonder if this is going to be a broader issue -- are media apps for older OS versions not going to work because they expect the blanket read permissions to be given?

grzesiek2010 commented:
> Nice. But does this mean that recording video with that default app would never work? I wonder if this is going to be a broader issue -- are media apps for older OS versions not going to work because they expect the blanket read permissions to be given?

I tested emulators with apis 22-30 and noticed that problem on api 23 and 25. In those cases recording is not possible with default camera apps. Granting storage permission fixes the issue but I thought it was a bug in those camera apps. In the doc where recording videos is described https://developer.android.com/training/camera/videobasics there is nothing about storage permission and we use exactly that way.

Apparently they love to confuse people.

I also read https://developer.android.com/training/data-storage/shared/media and https://developer.android.com/training/data-storage/shared/documents-files and my understanding is that we shouldn't need that permissions.

grzesiek2010 committed:
Improved isReadUriPermissionGranted() method

mmarciniak90 commented:
**Tested with success**

Verified on Android 7.0, 8.1, 10.0

Verified cases:
- Image widgets
   - choose image from Image directory
   - choose image from Downloads
   - choose image from Google Drive
   - take new photo
   - Camera permissions
- Audio widget
   - choose audio file from Audio directory
   - choose audio file from Downloads
   - choose audio file from Google Drive
   - record new audio
   - Microphone permissions
- Video widget
   - choose video file from Video directory
   - choose video file from Downloads
   - choose video file from Google Drive
   - record new video
   - Camera and Microphone permissions
- File widget
   - choose any file for file widget from Images/Videos/Audio
   - choose any file for file widget from Downloads
   - choose any file for file widget from Google Drive
   - verified file types: amr, ogg, m4a, mp3, 3gp, gif, jpeg, jpg, png, pdf, doc, docx, txt, xls, xlsx, csv, xml, rtf, zip, rar files
- Location permissions
- Contact permissions
- Telephone permissions
- enabled storage permission manually
- fresh installation
- import settings using a qr code file
- import settings from .settings / json file
- custom splash screen images - file saved in /storage/emulated/0/Android/data/org.odk.collect.android/files 
- upgrade when storage permission is enabled
- upgrade when storage permission is disabled
   - if files are migrated, storage permissions are not required
   - if files are not migrated, storage permissions are required
- jpg with the correct MIME type but no extension (internal storage and GD) - verified on images widgets and file widget
- text file with extension .fake and the correct MIME type (internal storage and GD) - verified on file widget

kkrawczyk123 commented:
Verified also on Androids 5.1, 7.1 and 9.0.

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"

grzesiek2010 commented:
Regarding that problems with permissions: If it's not our fault bug external apps that do not grant uri permission we may need to consider adding something like: https://github.com/getodk/collect/commit/17431bbc5d663f01a349c43d40682674524dfc31
